### ISSUE TITLE
From fully connected to fully convolutional in maskrcnn

### ISSUE THREAD

**Issue by John1231983** [matterport/Mask_RCNN#286]
Hello all, I have to work with the arbitrary size of the testing image without using resize it to a fixed dimension. In the training, I randomly crop the image into `256x256` to obtain a trained model. After that, I will test the model with different size in inference phase. 

To do it, I followed the FCN paper that replaces the last fully connected layer to 1x1 convolutional layer. Look inside the implementation, I found two lines used fully connected layer
[Line 920](https://github.com/matterport/Mask_RCNN/blob/master/model.py#L920)
```
mrcnn_class_logits = KL.TimeDistributed(KL.Dense(num_classes),
                                            name='mrcnn_class_logits')(shared)
```
[Line 927](https://github.com/matterport/Mask_RCNN/blob/master/model.py#L927)
```
x = KL.TimeDistributed(KL.Dense(num_classes * 4, activation='linear'),
                           name='mrcnn_bbox_fc')(shared)
```

To convert from fully connected to fully convolutional layer, Is it right if I change these above lines? Could you give me some advice?
This is my current solution but it does not work

   ```
 mrcnn_class_logits = KL.TimeDistributed(KL.Conv2D(num_classes, (1, 1)),
                                            name='mrcnn_class_logits')(x)
    mrcnn_probs = KL.TimeDistributed(KL.Activation("softmax"),
                                     name="mrcnn_class")(mrcnn_class_logits)
    # BBox head
    # [batch, boxes, num_classes * (dy, dx, log(dh), log(dw))]
    x = KL.TimeDistributed(KL.Conv2D(num_classes*4, (1, 1)),
                                            name='mrcnn_bbox_fc')(x)
```

**Comment by Superlee506** [comment]
@John1231983 It is already a FCN model,  please refer to[ Line 906-917](https://github.com/matterport/Mask_RCNN/blob/master/model.py#L906)

**Comment by John1231983** [comment]
Thanks, Superlee506, but why it need to resize to same size with training size in the inference phase? Refer: [L1193](https://github.com/matterport/Mask_RCNN/blob/master/model.py#L1193) and [L2260](https://github.com/matterport/Mask_RCNN/blob/master/model.py#L2260). 
I asked this question because I trained in input sizes of 128x128 but when I test with size of 512x512, it shows the error about mismatching size. Note that, I have deleted the above resize function

**Comment by FruVirus** [comment]
@John1231983,

When you trained with the input sizes of 128 x 128, did you allow the network to resize them to 1024 x 1024 during the training?

**Comment by John1231983** [comment]
@FruVirus: No. I  did not resize it to 1024x1024 because resize will add zero padding that may loss the image statistic. In original implementation, the author resize it to fixed dimension such as 1024x1024 by zero padding from arbitrary input size. While, my code ramdomly crop the image to a fixed size 128x128. I can train and inference successfully with my 128x128 but I am not able to run with other size.

**Comment by FruVirus** [comment]
@John1231983 , this is my guess:

The reason why you are able to successfully train and inference using the same size image (i.e., 128 x 128) is because the spatial dimensions of the FPN feature maps that you generate at each level during training depends on the input image spatial dimensions. If you inference using same size images, then there is no dimension mismatch.

However, if you inference with a different dimension image, then the spatial dimensions of the FPN feature maps you trained using 128 x 128 images won't match those created by using 512 x 512 images. In this case, I don't think the weight matrices would match either.

**Comment by John1231983** [comment]
Right. That is my purpose of this question. I try to make FPN works in any size by use fully convolution instead of fully connected . Is it possible?

**Comment by FruVirus** [comment]
@John1231983 , gotcha.

So when you forward a 512 x 512 image, you'll get feature maps with spatial dimensions that are larger than those you would get by forwarding a 128 x 128 image. Perhaps you can then try down-sampling the larger feature maps to the same spatial dimensions as the feature maps of the 128 x 128 image.

**Comment by John1231983** [comment]
Good idea. So I will down sampling it in the FPN function, right? Or after resnet50 ?  I think resize feature is more correct in case

**Comment by FruVirus** [comment]
@John1231983 

The first place I would resize is in resnet_graph(). I think if you ensured that the output spatial sizes of each ResNet stage is the same after the call to resnet_graph(), then it'll also take care of the FPN portion. Taking a look through the rest of the code, the calls to fpn_classifier_graph() and build_fpn_mask_graph() during inference should also work since the spatial dimensions of mrcnn_feature_maps will be fixed.

**Comment by CMCDragonkai** [comment]
If it's possible to turn MaskRCNN to accept images of any sizes (without the initial resizing + padding operation), then what would the anchor size scales `RPN_ANCHOR_SCALES` and `RPN_ANCHOR_RATIOS` be? They would have to be based on the sizes of objects you expect to find?

**Comment by JonathanCMitchell** [comment]
@John1231983 In your code inside the `pyramidROIAlign()` [L365](https://github.com/John1231983/kaggle-ds-bowl-2018-baseline/blob/master/model.py#L365) call it seems that your ROI level will not reach P4 or P5 of the feature pyramid network if your image resolution is 128, 128. Are you experiencing issues with this or did the change to your `ANCHOR_SCALES` [L25](https://github.com/John1231983/kaggle-ds-bowl-2018-baseline/blob/master/bowl_config.py#L25) solve this for you?

**Comment by waleedka** [comment]
This mode is now supported in the latest update on the master branch. Only for inferencing, though. For training, the network still requires that all images are resized to the same size so they can be put in a batch. 

You can enable it using `config.IMAGE_RESIZE_MODE`:

* For training, use `IMAGE_RESIZE_MODE = "square"`. This will scale and pad the images such that they're in a box of size `IMAGE_MAX_DIM x IMAGE_MAX_DIM`.
* For inferencing, you can use `square` as well, which is the default behavior. But now you can also use the new `pad64` which uses the image in whichever size you give it. If you also provide IMAGE_MIN_DIM, then before padding it scales up the image (if needed) such that the small side is = IMAGE_MIN_DIM.

**Comment by CMCDragonkai** [comment]
What does 64 in pad64 mean?

**Comment by waleedka** [comment]
It's in the comments in **config.py**:
```
    # pad64:  Pads width and height with zeros to make them multiples of 64.
    #         If IMAGE_MIN_DIM is not None, then scale the small side to
    #         that size before padding. IMAGE_MAX_DIM is ignored in this mode.
    #         The multiple of 64 is needed to ensure smooth scaling of feature
    #         maps up and down the 6 levels of the FPN pyramid (2**6=64).
```
